Binding System refactor

grokys opened the pull request:
## What does the pull request do?

A major refactor of the binding system:

- Simplified the `BindingExpression` code: no more separate `BindingObserver` and `BindingExpression` - everything is now implemented in `BindingExpression` itself
- Added a new `BindingExpressionBase` that acts as a base-class for the various types of binding expressions
- Added a new `Bind` method on `AvaloniaObject` that returns the created `BindingExpressionBase`
- `BindingExpression`s can now be put directly into the `ValueStore` meaning that no intermediate wrapper objects are required, reducing memory usage
- Because `BindingExpression` now has more information about what's being bound, we can log less extraneous warnings (#9422, #5762)

This allows us to add a new feature in this PR and more new features in future:

- Added `UpdateSourceTrigger` (#3754)
- Added `BindingExpressionBase.UpdateSource` and `UpdateTarget`

## Benchmarks

### `master`

|                                    Method |     Mean |   Error |  StdDev |    Gen0 | Allocated |
|------------------------------------------ |---------:|--------:|--------:|--------:|----------:|
| Setup_DataContext_Property_Binding_OneWay | 127.5 us | 1.07 us | 0.95 us | 12.9395 | 212.63 KB |
| Setup_DataContext_Property_Binding_TwoWay | 167.9 us | 1.12 us | 1.05 us | 17.0898 | 280.59 KB |

|                                            Method |      Mean |     Error |    StdDev |   Gen0 | Allocated |
|-------------------------------------------------- |----------:|----------:|----------:|-------:|----------:|
| Produce_DataContext_Property_Binding_Value_OneWay |  1.454 us | 0.0167 us | 0.0140 us | 0.1450 |   2.38 KB |
| Produce_DataContext_Property_Binding_Value_TwoWay | 75.302 us | 0.3877 us | 0.3627 us | 2.0752 |  35.87 KB |

|                       Method |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|----------------------------- |---------:|---------:|---------:|-------:|----------:|
| Setup_TemplateBinding_OneWay | 45.46 us | 0.292 us | 0.274 us | 4.8828 |  80.47 KB |
| Setup_TemplateBinding_TwoWay | 84.09 us | 0.315 us | 0.279 us | 7.8125 | 128.13 KB |

|                               Method |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|------------------------------------- |---------:|---------:|---------:|-------:|----------:|
| Produce_TemplateBinding_Value_OneWay | 24.76 us | 0.085 us | 0.075 us | 1.1292 |  18.77 KB |
| Produce_TemplateBinding_Value_TwoWay | 65.14 us | 0.259 us | 0.242 us | 2.1973 |  37.22 KB |

### `refactor/bindingexpressions-in-valuestore`

|                                    Method |     Mean |   Error |  StdDev |   Gen0 | Allocated |
|------------------------------------------ |---------:|--------:|--------:|-------:|----------:|
| Setup_DataContext_Property_Binding_OneWay | 122.8 us | 0.70 us | 0.62 us | 5.6152 |  95.45 KB |
| Setup_DataContext_Property_Binding_TwoWay | 130.3 us | 0.92 us | 0.81 us | 7.0801 |  118.1 KB |

|                                            Method |      Mean |     Error |    StdDev |   Gen0 | Allocated |
|-------------------------------------------------- |----------:|----------:|----------:|-------:|----------:|
| Produce_DataContext_Property_Binding_Value_OneWay |  1.457 us | 0.0073 us | 0.0069 us | 0.0725 |   1.21 KB |
| Produce_DataContext_Property_Binding_Value_TwoWay | 30.309 us | 0.1168 us | 0.1036 us | 1.0681 |  17.84 KB |

|                       Method |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|----------------------------- |---------:|---------:|---------:|-------:|----------:|
| Setup_TemplateBinding_OneWay | 49.83 us | 0.266 us | 0.248 us | 4.2725 |  70.45 KB |
| Setup_TemplateBinding_TwoWay | 48.76 us | 0.216 us | 0.202 us | 4.2725 |  70.45 KB |

|                               Method |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|------------------------------------- |---------:|---------:|---------:|-------:|----------:|
| Produce_TemplateBinding_Value_OneWay | 22.19 us | 0.198 us | 0.185 us | 1.4648 |  24.27 KB |
| Produce_TemplateBinding_Value_TwoWay | 55.34 us | 0.169 us | 0.158 us | 2.8687 |  47.71 KB |

As can be seen, almost all of the benchmarks have significantly better memory usage and most have better or equal performance. There are a couple of exceptions though: `Setup_TemplateBinding_OneWay` and `Produce_TemplateBinding_Value_OneWay`.

What's happening here is that we're now storing the current value from the binding as a boxed value in `TemplateBinding` whereas before an `IValueEntry<T>` was being created for each template binding, meaning that the value didn't need to be boxed. `Setup_TemplateBinding_OneWay` being slightly slower is a result of this: a little more GC is taking place. This is offset by the reduced memory cost for setting up a `TemplateBinding` after the refactor.

## Breaking changes

- `Binding.Source` default value is now `Unsetvalue` - setting it to null now means "bind to null source" instead of "bind to data context" (as in WPF)
- `FallbackValue` and `TargetNullValue` conversion errors are no longer passed to the binding target as `BindingNotification`s - just logs an error
- `TemplateBinding` now throws if Mode is set to `OneTime` or `OneWayToSource` - previously it silently reverted to `OneWay`

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues

Fixes #3754 
Fixes #5762 
Fixes #9422 
Fixes #10469
Fixes #10793
Fixes #13533
Fixes #13734

grokys committed:
Initial refactor of binding infrastructure.

- `ExpressionObserver` has been removed and its functionality merged with `BindingExpression`
- `BindingExpression` handles all types of `BindingMode` itself; doesn't require `BindingOperations.Apply` to set up a separate observable for `TwoWay/`OneWayToSource` bindings
  - This allows us to fix some long-standing issues with `OneWayToSource` bindings
- Expression nodes have been refactored
  - No longer split between `Avalonia.Base` and `Avalonia.Markup`
  - Categorize them according to whether they use reflection or not

A few tests are failing around binding warnings: this is because the next step here is to fix binding warnings.

grokys committed:
Make default binding Source = UnsetProperty.

Null is a theoretically valid value for `Source`; setting it to null shouldn't mean "use the data context".

grokys committed:
Move logging to BindingExpression.

As `BindingExpression` now has enough information to decide when it's appropriate to log an error/warning or not.

Fixes #5762
Fixes #9422

grokys committed:
Update ncrunch config.

grokys committed:
WIP: Benchmarks

grokys committed:
Add compatibility hack for older compiled bindings.

Previously, `CompiledBindingPathBuilder` didn't have a `TemplatedParent` method and instead the XAML compiler rewrite templated parent bindings to be a `$self.TemplateParent` property binding. resulting in extraneous logs.

Add a constructor with an  `apiVersion`  to `CompiledBindingPathBuilder` which will be used by newer versions of the XAML compiler, and if a usage is detected using an `apiVersion` of 0, then upgrade `$self.TemplatedParent` to use a `TemplatedParentPathElement`.

grokys committed:
Log errors from property accessors.

grokys committed:
Don't log errors for named control bindings...

...on elements which aren't yet rooted.

grokys committed:
Log errors for failed conversions.

grokys committed:
Use consistent wording for binding warnings.

"Could not convert" instead of "Cannot convert".

grokys committed:
Log warnings for converter exceptions.

grokys committed:
Don't convert new TargetTypeConverters each time.

grokys committed:
Added failing test for implicit conversion.

grokys committed:
Support cast operators in compiled bindings.

A bit of a hack as we'd ideally not be using reflection when using compiled bindings.

grokys committed:
This shouldn't be a public API.

Should only be used for tests.

grokys committed:
Make enum/int conversion work.

grokys committed:
Check for SetValue equality after conversion.

And also use "identity equals" where value types and strings use `object.Equals` and reference types use `object.ReferenceEquals`.

grokys committed:
Added ConverterCulture back to bindings.

grokys committed:
Merge branch 'master' into refactor/bindings

grokys committed:
Use BindingExpression directly in ValueStoe.

grokys committed:
Fix merge error.

Removed deleted files from csproj that were re-added due to indentation changes.

grokys committed:
Introduce BindingExpressionBase.

And `UntypedBindingExpressionBase`.

grokys committed:
Make TemplateBinding a BindingExpression.

grokys committed:
Make DynamicResource use a BindingExpression.

grokys committed:
WIP: Start exposing a BindingExpression API.

grokys committed:
Finish exposing a BindingExpression API.

grokys committed:
Fix OneTimeBinding.

grokys committed:
Remove unneeded classes/methods.

grokys committed:
Don't call obsolete API.

grokys committed:
Make BindingExpressionBase the public API.

This matches WPF's API.

grokys committed:
Added BindingExpressionBase.UpdateTarget.

grokys committed:
Initial implementation of UpdateSourceTrigger.

grokys committed:
Don't use weak references for values.

If they're boxed values, they can get collected.

grokys committed:
No need for virtual/generic methods here now.

grokys committed:
Reintroduce support for binding anchors.

Turns out these were needed by animations, just our animation system has no unit tests so I missed that fact earlier. Add a basic animation unit test that fails without anchor support, and add binding anchors back in. Currently a private API as I suspect this feature shouldn't be needed outside the framework.

grokys committed:
Include new property in clone.

And add real-life example of `UpdateSourceTrigger=LostFocus` to BindingDemo.

grokys committed:
Merge branch 'master' into refactor/bindings

grokys committed:
Merge branch 'master' into refactor/bindings

grokys committed:
Fix merge error.

grokys committed:
Updated BindingExpression tests.

- Make them run for both compiled and reflection bindings (found a bunch of tests that fail with compiled bindings)
- Make them not depend on converting the `BindingExpression` to an observable and instead test the end result of the binding on an `AvaloniaObject`

grokys committed:
Fix compiled binding indexer tests.

grokys committed:
Use data validation plugins in PropertyAccessorNode.

Added a warning suppression for now: we may need a separate `DataValidators` list for AOT-friendly plugins.

grokys committed:
Don't separate plugins by reflection.

`DataAnnotationsValidationPlugin` is public and so it can't be moved. No point in moving the others if this one will be in the wrong place.

grokys committed:
Remove unneeded methods.

grokys committed:
Make reflection binding tests use a string.

Convert the `System.Linq.Expression` to a string and then use this, as reflection bindings will always be instanced with a string path.

grokys committed:
Merge branch 'master' into refactor/bindings

TomEdwardsEnscape commented on src/Avalonia.Base/Data/Core/BindingExpression.cs:
```diff
@@ -1,389 +1,549 @@
+/// </remarks>
+internal partial class BindingExpression : UntypedBindingExpressionBase, IDescription, IDisposable
 {
+    private readonly WeakReference<object?>? _source;
```
Very happy to see this PR!

```suggestion
    ~BindingExpression()
    {
        _source.Free(); // field must not be readonly: this call mutates its value
    }
    
    private GCHandle _source;
```

For private fields, it's reasonable to use `GCHandle` directly and replace the `WeakReference` finalizer with our own. This eliminates one object allocation per binding expression.

grokys replied:
Probably worth doing in a separate PR so it can be reverted easily if it causes a problem.

maxkatz6 commented on src/Avalonia.Base/Data/Core/IBinding2.cs:
```diff
@@ -0,0 +1,9 @@
+
+internal interface IBinding2 : IBinding
+{
+    BindingExpressionBase Instance(
```
@grokys do I understood it correctly, that InstancedBinding is now obsolete with new bindings, as BindingExpression works as a specific instance of a binding?

And I wonder what's a plan with dealing IBinding2.
In general, it looks like you can combine it into a single interface (since IBinding is not client implementable). But keep old `Instance` method with Obsolete attribute.

grokys replied:
Yeah `IBinding2` is a total hack, ideally it'd all be implemented as internal code in a `BindingBase` class but too late for that now. I need to add some comments on how it should be refactored for 12.0.

robloo replied:
Just to be clear, this type of thing will be cleaned up for 12.0 as a breaking change, right?

timunie replied:
we should add later on comments `// TODO12: Remove me` otherwise we may miss it.

workgroupengineering commented:
With this PR is it possible to obtain the BindingExpression linked to an AvaloniaProperty?

I would like to add this information to DevTools.

grokys commented:
> With this PR is it possible to obtain the BindingExpression linked to an AvaloniaProperty?

Not yet, though that is planned!

timunie commented:
List of potential solved issues: 

## updatesourcetrigger 

- #13734
- #13533
- #10469
- #3754
- #1362

## BindingBase / IBinding
- #10956
- #13086
- #3815
- #3038
- #3395
- #2959

## Other
- #4466
- #1326 
- #3193

robloo commented:
Great work! It's an early gift for the holidays!

toomasz commented:
Good work! Can't wait to have it released

maxkatz6 commented on src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlBindingPathHelper.cs:
```diff
@@ -245,12 +246,18 @@ private static XamlIlBindingPathNode TransformBindingPath(AstTransformationConte
+
+                        if (templatedParentType is null)
+                        {
+                            throw new XamlParseException("A binding with a TemplatedParent RelativeSource has to be in a ControlTemplate.", lineInfo);
```
We should use as specific exception as possible.
```suggestion
                            throw new XamlTransformException("A binding with a TemplatedParent RelativeSource has to be in a ControlTemplate.", lineInfo);
```

grokys committed:
Added TODO12 plan for IBinding2.

grokys committed:
Use more specific exception.

grokys commented on src/Avalonia.Base/Data/Core/ExpressionNodes/ISettableNode.cs:
```diff
@@ -0,0 +1,23 @@
+    /// <param name="value">The value to write.</param>
+    /// <param name="nodes">The expression nodes in the binding.</param>
+    /// <returns>True if the value was written sucessfully; otherwise false.</returns>
+    bool WriteValueToSource(object? value, IReadOnlyList<ExpressionNode> nodes);
```
It's used in `LogicalNotNode`: https://github.com/AvaloniaUI/Avalonia/pull/13970/files#diff-be7f1b377e935591bfbad00a122c53d16491e780321fc39044dd2e6417794cacR26

grokys committed:
Fix nits from code review.

grokys committed:
Make expression nodes sealed where possible.

grokys commented on src/Avalonia.Base/Data/Core/ExpressionNodes/ArrayIndexerNode.cs:
Sure, I think that would make sense to add in a separate PR at this point.

grokys committed:
Merge branch 'master' into refactor/bindings

grokys commented on src/Avalonia.Base/Data/Core/BindingExpression.cs:
```diff
@@ -1,389 +1,549 @@
+    private readonly BindingMode _mode;
+    private readonly IReadOnlyList<ExpressionNode> _nodes;
+    private readonly TargetTypeConverter? _targetTypeConverter;
+    private readonly UncommonFields? _uncommon;
```
Done, though added the comment to the `UncommonFields` class itself.

agnauck commented:
Would it be possible to implement WPFs RelativeSource.PreviousData binding with this PR?
See: https://github.com/AvaloniaUI/Avalonia/issues/3733

grokys committed:
Unsubscribe on Stop, don't re-subscribe.

D'oh.

grokys committed:
Tweak ExpressionNode lists.

Saves a few K in benchmarks and it's a cleaner API.

grokys committed:
Add a pooled option in BindingExpressionGrammar.

Micro-optimization.

grokys committed:
Avoid allocations when enumerating binding plugins.

grokys committed:
Add IBinding2 support to observable bind overloads.

In the case of `TemplateBinding`, the `IObservable<object?>` bind overload is selected by C#. Add an explicit check for an `IBinding2` here to use the more performant code-path.

grokys committed:
Remove disposed binding from ImmediateBindingFrame.

grokys committed:
Added TemplateBinding benchmarks.

grokys commented:
Ok, added some benchmarks and I think this PR is ready for review/testing now.

Takoooooo commented on src/Markup/Avalonia.Markup.Xaml/Avalonia.Markup.Xaml.csproj:
```diff
@@ -8,58 +8,95 @@
+    <Compile Include="Converters\FontFamilyTypeConverter.cs" />
+    <Compile Include="Converters\TimeSpanTypeConverter.cs" />
+    <Compile Include="Extensions.cs" />
+    <Compile Include="MarkupExtension.cs" />
+    <Compile Include="MarkupExtensions\CompiledBindingExtension.cs" />
+    <Compile Include="MarkupExtensions\CompiledBindings\CompiledBindingPath.cs" />
+    <Compile Include="MarkupExtensions\CompiledBindings\PropertyInfoAccessorFactory.cs" />
+    <Compile Include="MarkupExtensions\CompiledBindings\TaskStreamPlugin.cs" />
+    <Compile Include="MarkupExtensions\DynamicResourceExtension.cs" />
+    <Compile Include="MarkupExtensions\On.cs" />
+    <Compile Include="MarkupExtensions\OnFormFactorExtension.cs" />
+    <Compile Include="MarkupExtensions\OnPlatformExtension.cs" />
+    <Compile Include="MarkupExtensions\ResolveByNameExtension.cs" />
+    <Compile Include="MarkupExtensions\StaticResourceExtension.cs" />
+    <Compile Include="Parsers\PropertyParser.cs" />
+    <Compile Include="Converters\BitmapTypeConverter.cs" />
+    <Compile Include="Converters\IconTypeConverter.cs" />
+    <Compile Include="Converters\AvaloniaPropertyTypeConverter.cs" />
+    <Compile Include="Converters\PointsListTypeConverter.cs" />
+    <Compile Include="MarkupExtensions\ReflectionBindingExtension.cs" />
+    <Compile Include="MarkupExtensions\RelativeSourceExtension.cs" />
+    <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="RuntimeXamlLoaderConfiguration.cs" />
+    <Compile Include="RuntimeXamlLoaderDocument.cs" />
+    <Compile Include="Styling\MergeResourceInclude.cs" />
+    <Compile Include="Styling\ResourceInclude.cs" />
+    <Compile Include="Styling\StyleInclude.cs" />
+    <Compile Include="Templates\ControlTemplate.cs" />
+    <Compile Include="Templates\DataTemplate.cs" />
+    <Compile Include="Templates\FocusAdornerTemplate.cs" />
+    <Compile Include="Templates\ItemsPanelTemplate.cs" />
+    <Compile Include="Templates\Template.cs" />
+    <Compile Include="Templates\TemplateContent.cs" />
+    <Compile Include="Templates\TreeDataTemplate.cs" />
+    <Compile Include="XamlIl\Runtime\IAvaloniaXamlIlControlTemplateProvider.cs" />
+    <Compile Include="XamlIl\Runtime\IAvaloniaXamlIlParentStackProvider.cs" />
+    <Compile Include="XamlIl\Runtime\IAvaloniaXamlIlXmlNamespaceInfoProviderV1.cs" />
+    <Compile Include="XamlIl\Runtime\XamlIlRuntimeHelpers.cs" />
+    <Compile Include="XamlLoadException.cs" />
+    <Compile Include="XamlTypes.cs" />
```
I guess that it is incorrect that all files are included twice here now

grokys replied:
Must have been a merge error - well spotted! Fixed.

Takoooooo commented:
I am planning to go through all the related issues @timunie linked and validate whether they are fixed or not.
List of potential solved issues: 

## updatesourcetrigger 

- #13734
- #13533
- #10469
✅ #3754
❌ #1362 (That's a different issue. Not related to UpdateSourceTrigger so it isn't fixed)

## BindingBase / IBinding
- #10956
- #13086
- #3815
- #3038
- #3395
- #2959

## Other
- #4466
- #1326 
- #3193 
- #10793

timunie commented:
as pointed out in #14145 I get an exception trying this branch on samples > ControlCatalog > ColorPicker opening the control at least twice. 

![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/b8f14ffb-490f-491b-95a0-5ff51cd3e2ca)

grokys committed:
Remove duplicate items.

Seems to have been caused by a merge error.

grokys committed:
Fix exception when closing color picker.

And add tests.

grokys commented:
> as pointed out in https://github.com/AvaloniaUI/Avalonia/issues/14145 I get an exception trying this branch on samples > ControlCatalog > ColorPicker opening the control at least twice.

@timunie fixed!

timunie commented:
@billhenn one more crash was fixed just a little earlier. So probably check it next week again. Feedback is very much appreciated 👍

maxkatz6 commented on src/Avalonia.Base/AvaloniaObject.cs:
```diff
@@ -440,7 +455,16 @@ public void SetCurrentValue<T>(StyledProperty<T> property, T value)
             ValidatePriority(priority);
 
-            return _values.AddBinding(property, source, priority);
+            if (source is IBinding2 b)
+            {
+                if (b.Instance(this, property, null) is not UntypedBindingExpressionBase expression)
+                    throw new NotSupportedException("Binding returned unsupported IBindingExpression.");
+                return GetValueStore().AddBinding(property, expression);
+            }
```
Looks like priority is ignored in this code path. Is it expected?

grokys replied:
Fixed.

grokys committed:
Don't skip converter when binding to self.

Takoooooo commented:
#10956 Is not fixed. I think it may make sense to fix it or at least give a reasoning why it cant be fixed\improve an exception message 
![image](https://github.com/AvaloniaUI/Avalonia/assets/53405089/071587ca-15b4-45a8-a2f2-41a28c94a384)

grokys committed:
Don't pass UnsetValue to converters.

This follows WPF behavior.

timunie commented:
@grokys https://github.com/AvaloniaUI/Avalonia/issues/14221#issuecomment-1893316324

grokys committed:
Log element name if present.

More useful than just logging the control hash code.

grokys committed:
Respect binding priority.

grokys commented:
> https://github.com/AvaloniaUI/Avalonia/issues/10956 Is not fixed. I think it may make sense to fix it or at least give a reasoning why it cant be fixed\improve an exception message

Yeah this is a limitation of the reflection binding infrastructure, and no new features have been added to this in this PR. It should be addressed in a separate PR IMO as this PR already does too much.

billhenn commented:
The most recent PR build fixes all the issues we've encountered from earlier PR builds.  This is a fantastic update for Avalonia.

grokys committed:
Throw on mismatched binding priorities.

We don't want to respect the binding priority in this case as it breaks `TemplateBindings` when the default `LocalValue` priority is passed. Instead make sure that the priority parameter matches that of the expression.

This reverts commit a72765d70514506d94b39ba33a3d91e11cd6dcd4.

grokys committed:
Convert to target type in TemplateBinding.

grokys committed:
Short-circuit target type conversion for same types.

grokys committed:
Merge branch 'master' into refactor/bindings

Takoooooo reviewed (approved):
Tested on a few projects, LGTM!

rabbitism commented:
yeaaaah🚀

BAndysc commented:
@grokys Found a bug: after the refactor, OneWayToSource bindings don't work for read-only direct properties, i.e.:

```
<Border>
  <Border.ContextFlyout>
    <Flyout IsOpen="{CompiledBinding IsOpen, Mode=OneWayToSource}">
      <Border  />
    </Flyout>
  </Border.ContextFlyout>
</Border>
  ```
  
  ```
  Unhandled exception. System.ArgumentException: The property IsOpen is readonly.
   at Avalonia.DirectProperty`2.InvokeSetter(AvaloniaObject instance, BindingValue`1 value) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/DirectProperty.cs:line 119
   at Avalonia.AvaloniaObject.SetDirectValueUnchecked[T](DirectPropertyBase`1 property, BindingValue`1 value) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/AvaloniaObject.cs:line 819
   at Avalonia.DirectPropertyBase`1.RouteSetDirectValueUnchecked(AvaloniaObject o, Object value) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/DirectPropertyBase.cs:line 164
   at Avalonia.PropertyStore.ValueStore.Avalonia.Data.Core.IBindingExpressionSink.OnChanged(UntypedBindingExpressionBase instance, Boolean hasValueChanged, Boolean hasErrorChanged, Object value, BindingError error) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/PropertyStore/ValueStore.cs:line 765
   at Avalonia.Data.Core.UntypedBindingExpressionBase.PublishValue(Object value, BindingError error) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/Data/Core/UntypedBindingExpressionBase.cs:line 429
   at Avalonia.Data.Core.BindingExpression.StartCore() in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/Data/Core/BindingExpression.cs:line 379
   at Avalonia.Data.Core.UntypedBindingExpressionBase.Start(Boolean produceValue) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/Data/Core/UntypedBindingExpressionBase.cs:line 473
   at Avalonia.Data.Core.UntypedBindingExpressionBase.AttachAndStart(IBindingExpressionSink subscriber, AvaloniaObject target, AvaloniaProperty targetProperty, BindingPriority priority) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/Data/Core/UntypedBindingExpressionBase.cs:line 208
   at Avalonia.PropertyStore.ValueStore.AddBinding(AvaloniaProperty property, UntypedBindingExpressionBase source) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/PropertyStore/ValueStore.cs:line 55
   at Avalonia.AvaloniaObject.Bind(AvaloniaProperty property, IBinding binding, Object anchor) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/AvaloniaObject.cs:line 635
   at Avalonia.AvaloniaObject.Bind(AvaloniaProperty property, IBinding binding) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/AvaloniaObject.cs:line 421
   at Avalonia.AvaloniaObjectExtensions.Bind(AvaloniaObject target, AvaloniaProperty property, IBinding binding, Object anchor) in /Users/bartek/Programowanie/temp/Avalonia/src/Avalonia.Base/AvaloniaObjectExtensions.cs:line 251
   at ControlCatalog.MainWindow.!XamlIlPopulate(IServiceProvider, MainWindow) in /Users/bartek/Programowanie/temp/Avalonia/samples/ControlCatalog/MainWindow.xaml:line 42
   at ControlCatalog.MainWindow.!XamlIlPopulateTrampoline(MainWindow)
   at ControlCatalog.MainWindow.InitializeComponent() in /Users/bartek/Programowanie/temp/Avalonia/samples/ControlCatalog/MainWindow.xaml.cs:line 32
   at ControlCatalog.MainWindow..ctor() in /Users/bartek/Programowanie/temp/Avalonia/samples/ControlCatalog/MainWindow.xaml.cs:line 25
   at ControlCatalog.App.SetCatalogThemes(CatalogTheme theme) in /Users/bartek/Programowanie/temp/Avalonia/samples/ControlCatalog/App.xaml.cs:line 99
   at ControlCatalog.App.Initialize() in /Users/bartek/Programowanie/temp/Avalonia/samples/ControlCatalog/App.xaml.cs:line 40
```
